packages/jest-mock/src/index.ts

```diff
@@ -1003,31 +1022,38 @@ export class ModuleMocker {
     return fn;
   }
 
-  spyOn<T extends {}, M extends NonFunctionPropertyNames<T>>(
-    object: T,
-    methodName: M,
-    accessType: 'get',
-  ): SpyInstance<T[M], []>;
+  spyOn<
+    T extends ClassLike | ObjectLike,
+    M extends NonFunctionPropertyNames<T>,
+  >(object: T, methodName: M, accessType: 'get'): SpyInstance<T[M], []>;
 
-  spyOn<T extends {}, M extends NonFunctionPropertyNames<T>>(
+  spyOn<
+    T extends ClassLike | ObjectLike,
+    M extends NonFunctionPropertyNames<T>,
+  >(object: T, methodName: M, accessType: 'set'): SpyInstance<void, [T[M]]>;
+
+  spyOn<
+    T extends ClassLike | ObjectLike,
+    M extends ConstructorPropertyNames<Required<T>>,
+  >(
     object: T,
-    methodName: M,
-    accessType: 'set',
-  ): SpyInstance<void, [T[M]]>;
+    method: M,
+  ): T[M] extends new (...args: Array<any>) => any
+    ? SpyInstance<InstanceType<T[M]>, ConstructorArgsType<T[M]>>
+    : never;
```

Comment:
This is new overload. The rest of changes in overloads are just reformatting caused by the improved constrain.